samples/azure/azure-service-fabric-routing/Core_6/Shared/ReceiverSideDistribution/Pipeline/DistributeMessagesBasedOnHeader.cs

```diff
@@ -0,0 +1,49 @@
+using System;
+using System.Threading.Tasks;
+using NServiceBus;
+using NServiceBus.Pipeline;
+using NServiceBus.Transport;
+
+namespace Shared
+{
+    class DistributeMessagesBasedOnHeader : IBehavior<IIncomingPhysicalMessageContext, IIncomingPhysicalMessageContext>
+    {
+        readonly string localPartitionKey;
+        readonly Forwarder forwarder;
+
+        public DistributeMessagesBasedOnHeader(string localPartitionKey, Forwarder forwarder)
+        {
+            this.localPartitionKey = localPartitionKey;
+            this.forwarder = forwarder;
+        }
+
+        public Task Invoke(IIncomingPhysicalMessageContext context, Func<IIncomingPhysicalMessageContext, Task> next)
+        {
+            var intent = context.Message.GetMesssageIntent();
+            var isSubscriptionMessage = intent == MessageIntentEnum.Subscribe || intent == MessageIntentEnum.Unsubscribe;
+            var isReply = intent == MessageIntentEnum.Reply;
+
+            if (isSubscriptionMessage || isReply)
+            {
+                return next(context);
+            }
+
+            string messagePartitionKey;
+            var hasPartitionKeyHeader = context.MessageHeaders.TryGetValue(PartitionHeaders.PartitionKey, out messagePartitionKey);
+
+            //1. The header value isn't present (logical behavior will check message contents)
+            //2. The header value matches local partition key
+            if (!hasPartitionKeyHeader || messagePartitionKey == localPartitionKey)
+            {
+                return next(context);
+            }
+            context.Message.Headers[PartitionHeaders.PartitionKey] = messagePartitionKey;
```

Comment:
But line 40 is executed only if the header is there (and has a value different from local partition)